go/pkg/tree/tree_test.go

```diff
@@ -765,3 +774,249 @@ func TestFlattenTreeRepeated(t *testing.T) {
+		wantTreeChildren []*repb.Directory
+		wantCacheCalls   map[string]int
+	}{
+		{
+			desc: "Two files",
+			input: []*inputPath{
+				{path: "a/b/fooDir/foo", fileContents: fooBlob},
+				{path: "a/b/fooDir/bar", fileContents: barBlob},
+			},
+			wantBlobs:    [][]byte{fooBlob, barBlob},
+			wantTreeRoot: foobarDir,
+			wantCacheCalls: map[string]int{
+				"a/b/fooDir":     2,
+				"a/b/fooDir/bar": 1,
+				"a/b/fooDir/foo": 1,
+			},
+		},
+		{
+			desc: "Duplicate file contents",
+			input: []*inputPath{
+				{path: "a/b/fooDir/foo", fileContents: fooBlob},
+				{path: "a/b/fooDir/bar", fileContents: fooBlob},
+			},
+			wantBlobs: [][]byte{fooBlob, fooBlob},
+			wantTreeRoot: &repb.Directory{Files: []*repb.FileNode{
+				{Name: "bar", Digest: fooDgPb, IsExecutable: true},
+				{Name: "foo", Digest: fooDgPb, IsExecutable: true},
+			}},
+			wantCacheCalls: map[string]int{
+				"a/b/fooDir":     2,
+				"a/b/fooDir/bar": 1,
+				"a/b/fooDir/foo": 1,
+			},
+		},
+		{
+			desc: "Duplicate subdirectories",
+			input: []*inputPath{
+				{path: "a/b/fooDir/dir1/foo", fileContents: fooBlob},
+				{path: "a/b/fooDir/dir2/foo", fileContents: fooBlob},
+			},
+			wantBlobs: [][]byte{fooBlob},
+			wantTreeRoot: &repb.Directory{Directories: []*repb.DirectoryNode{
+				{Name: "dir1", Digest: fooDirDgPb},
+				{Name: "dir2", Digest: fooDirDgPb},
+			}},
+			wantTreeChildren: []*repb.Directory{fooDir},
+			wantCacheCalls: map[string]int{
+				"a/b/fooDir":          2,
+				"a/b/fooDir/dir1":     1,
+				"a/b/fooDir/dir1/foo": 1,
+				"a/b/fooDir/dir2":     1,
+				"a/b/fooDir/dir2/foo": 1,
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		root, err := ioutil.TempDir("", tc.desc)
+		if err != nil {
+			t.Fatalf("failed to make temp dir: %v", err)
+		}
+		defer os.RemoveAll(root)
+		if err := construct(root, tc.input); err != nil {
+			t.Fatalf("failed to construct input dir structure: %v", err)
+		}
+
+		t.Run(tc.desc, func(t *testing.T) {
+			wantBlobs := make(map[digest.Digest][]byte)
+			for _, b := range tc.wantBlobs {
+				wantBlobs[digest.NewFromBlob(b)] = b
+			}
+
+			gotBlobs := make(map[digest.Digest][]byte)
+			cache := newCallCountingMetadataCache(root, t)
+			chunkers, gotResult, err := ComputeOutputsToUpload(root, []string{"a/b/fooDir"}, chunker.DefaultChunkSize, cache)
+			if err != nil {
+				t.Fatalf("ComputeOutputsToUpload(...) = gave error %v, want success", err)
+			}
+			for _, ch := range chunkers {
+				blob, err := ch.FullData()
+				if err != nil {
+					t.Errorf("chunker %v FullData() returned error %v", ch, err)
+				}
+				gotBlobs[ch.Digest()] = blob
+			}
+			if diff := cmp.Diff(tc.wantCacheCalls, cache.calls, cmpopts.EquateEmpty()); diff != "" {
+				t.Errorf("ComputeOutputsToUpload(...) gave diff on file metadata cache access (-want +got) on blobs:\n%s", diff)
+			}
+			if len(gotResult.OutputDirectories) != 1 {
+				t.Fatalf("ComputeOutputsToUpload(...) expected result with an output directory, got %v", gotResult)
```

Comment:
nit: %+v?

Reply:
Oh! I didn't know about that one, thank you!